### PR TITLE
[WIP] DistributedDataParallel with FP8-support

### PR DESCRIPTION
**Pull request by wkcn**
**Description**
In this PR, the native pytorch `DistributedDataParallel` is overrided by `msmap.nn.parallel.distributed.DistributedDataParallel`, which supports ScalingTensor as the parameters. The weight gradients will be put in a bucket then all-reduced. The function `optimizer.all_reduce_grads` is deprecated.

**Major Revision**
- Add `msmap.nn.parallel.distributed.DistributedDataParallel`

**TODO**
- [x] asynchronous all reduce
- [ ] refactor code
- [ ] check training with deepspeed
- [ ] unit test for `msmap.nn.parallel.distributed.DistributedDataParallel`
- [ ] update ms-amp examples
     remove `optimizer.all_reduce_grads()`

### CODE REVIEW

File: msamp/nn/parallel/distributed.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        with torch.cuda.stream(self.reduction_stream):
+            # step 1: collect the gradients
+            param_ids = self.bucket_to_param_ids[bucket_id]
+            params = [self.parameters[i] for i in param_ids]
+            grads = [p.grad for p in params]
+            metas = [g.meta for g in grads]
+
+            # step 2: synchronize the amax
+            # shape of amaxs: (n, )
+            amaxs = torch.stack([g.max().float() for g in grads])
```
Comment (wkcn) [inline msamp/nn/parallel/distributed.py:107]: .abs().max()

**Comment by wkcn** [comment]
Close it because of the new PR #93
